<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="EDGE : EDGE bioinformatics">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
    <script src="javascripts/snowfall.jquery.js"></script>

    <title>EDGE</title>
  </head>

  <body>
  <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-70354817-3', 'auto');
  ga('send', 'pageview');
  
  $(document).ready(function(){
	$("a").attr('target', '_blank');
	$("#header_wrap").snowfall({deviceorientation : true, round : true, minSize: 1, maxSize:8,  flakeCount : 35});
  });
  </script>
    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/LANL-Bioinformatics/EDGE">View on GitHub</a>

          <h1 id="project_title">EDGE</h1>
          <h2 id="project_tagline">Empowering the Development of Genomics Expertise</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/LANL-Bioinformatics/EDGE/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/LANL-Bioinformatics/EDGE/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="what-is-edge" class="anchor" href="#what-is-edge" aria-hidden="true"><span class="octicon octicon-link"></span></a>What is EDGE?</h3>

<p>
EDGE bioinformatics was <strong>developed to help biologists process Next Generation Sequencing data</strong> (in the form of <strong>raw FASTQ</strong> files), even if they have little to no bioinformatics expertise. EDGE is a highly integrated and interactive web-based platform that is capable of running many of the standard analyses that biologists require for viral, bacterial/archaeal, and metagenomic samples. EDGE provides the following analytical workflows: <strong>quality trimming and host removal, assembly and annotation, comparisons against known references, taxonomy classification of reads and contigs, whole genome SNP-based phylogenetic analysis, and PCR analysis</strong>. EDGE provides an intuitive web-based interface for user input, allows users to visualize and interact with selected results (e.g. JBrowse genome browser), and generates a final detailed PDF report. Results in the form of tables, text files, graphic files, and PDFs can be downloaded. A user management system allows tracking of an individual’s EDGE runs, along with the ability to share, post publicly, delete, or archive their results.
</p>
<p>
While the design of EDGE was intentionally done to be as simple as possible for the user, there is still no single ‘tool’ or algorithm that fits all use-cases in the bioinformatics field. Our intent is to provide a detailed panoramic view of your sample from various analytical standpoints, but users are encouraged to have some insight into how each tool or workflow functions, and how the results should best be interpreted.
</p>	
<h3><a id="news" class="anchor" href="#news" aria-hidden="true"><span class="octicon octicon-link"></span></a>What's New? </h3>
<p>
  Updates 12/05/2016</b>:
  <ul>
    <li>EDGE version <a href="https://github.com/LANL-Bioinformatics/EDGE/releases"> v1.5.1 pre-release.</a>
    <li>NCBI is <a href="https://www.ncbi.nlm.nih.gov/home/develop/https-guidance.shtml">moving from http to https</a> for security and privacy improvement which caused SRA download feature failed. We have <a href="https://github.com/LANL-Bioinformatics/EDGE/blob/3e799f2b8f2bcd5852560d9574ade3cf167bc7bf/scripts/sra2fastq.pl">updated our script</a> to accommondate the change. In addition, if it failed to download from <a href="https://www.ncbi.nlm.nih.gov/sra">NCBI-SRA</a>, it will try <a href="http://www.ebi.ac.uk/ena">EBI-ENA</a> and <a href="https://trace.ddbj.nig.ac.jp/dra/index.html">DDBJ</a>.  </li>
    <li>One of the <a href="https://github.com/seninp-bioinfo/MetaComp">MetaComp</a> requirements udpated its version (cowplot) broken the EDGE installation: update R version from 3.2.2 to 3.3.2 and MetaComp from v1.1 to v1.2</li>
  </ul>
</p>
<p>
  Updates 11/29/2016</b>:
  <ul>
    <li>EDGE manuscript has been published on <em>Nucleic Acids Research</em> (2016). <a href="http://nar.oxfordjournals.org/content/early/2016/11/28/nar.gkw1027.abstract">See Abstract.</a></li>
    <li><a href="http://phys.org/news/2016-11-edge-bioinformatics-genomics.html">EDGE news on Phys.org</a></li>
  </ul>
</p>
<p>
  Updates 10/04/2016</b>:
  <ul>
    <li>EDGE version <a href="https://github.com/LANL-Bioinformatics/EDGE/releases"> v1.5.0 release.</a>
    <li>The EDGE documentation has branches with version <a href="https://edge.readthedocs.io/en/v1.5/"> v1.5 </a> and <a href="https://edge.readthedocs.io/en/latest/"> latest </a> for development (active updating)</li>
  </ul>
</p>
<p>
<p>
  Updates 06/08/2016</b>:
  <ul>
    <li>EDGE development branch on <a href="https://github.com/LANL-Bioinformatics/EDGE/tree/devel">GitHub.</a> and its <a href="https://figshare.com/articles/EDGE-poster-SFAF_pdf/3406246">Poster</a></li>
    <li>The EDGE documentation has branches with version <a href="https://edge.readthedocs.io/en/v1.1/"> v1.1 </a> and <a href="https://edge.readthedocs.io/en/latest/"> latest </a> for development (active updating)</li>
  </ul>
</p>
<p>
  Updates 03/08/2016</b>:
  <ul>
    <li><a href="https://github.com/LANL-Bioinformatics/EDGE/commit/fca9f97ca27d8d9f3541ecd5eaa4ac6b283f086b">fix bugs: batch submit point to wrong path. Escape symbol characters.</a></li>
    <li><a href="https://github.com/LANL-Bioinformatics/EDGE/commit/64a9618b76a15764021b420f49c34de635c80611">fix bug: run the EDGE through command line with absoulte path.</a></li>
    <li>remove unsupport taxanomy profiling tool in the example config file</li>
    <li>make the example apache configuration file ready for copy with EDGE installation script</li>
    <li><a href="http://biorxiv.org/content/early/2016/02/21/040477">A preprint manuscript is on bioRxiv</a>.</li>
  </ul>
</p>
<p>
  Update 11/09/2015</b>:
  <ul><li><a href="https://edge.readthedocs.org/en/v1.1/installation.html#edge-vmware-ovf-image">EDGE Virtual Machine Image</a></li></ul>
  Updates 10/15/2015</b>:
  <ul>
    <li>Version 1.1 source code released</li>
    <li>Have Docker image ready. See <a href="https://hub.docker.com/r/chienchilo/bioedge/">DOCKER HUB</a></li>
    <li>Databases can be downloaded through <a href="ftp://ftp.lanl.gov/public/genome/EDGE/">FTP server</a></li>
  </ul>
</p>

<h3><a id="using-edge" class="anchor" href="#using-edge" aria-hidden="true"><span class="octicon octicon-link"></span></a>Using EDGE</h3>
<p>
A demonstration version of EDGE is currently available at <a href="https://bioedge.lanl.gov/edge_ui/">https://bioedge.lanl.gov/edge_ui/</a> with example data sets available to the public to view and/or re-run and also allows EDGE runs of SRA/ENA data. (This webserver <strong>does not currently support upload of data due in part to LANL security regulations</strong>, however local installations are meant to be fully functional.) 

To run EDGE, go to the <a href="https://bioedge.lanl.gov/edge_ui/">EDGE webpage</a>. You will need to create an account in order to start a project, which you can do with the icon in the top right corner. Once you've created an account and have logged in, you can start a project. From the menu on the left, click "Run EDGE". This will open a page titled "Input your sample", and from here you will see a list of drop down menus that will allow you to label your project, input FASTQ files, and select the tools you would like to use. Pre-Processing, Assembly and Annotation, and Taxonomy Classification are turned on by default. 

While your project is running you can watch it's progress and system usage using the Job Progress sidebar, which opens using the icon in the top right corner. 

For those that want to run EDGE offline, installation instructions can be found in our documentation, which can be accessed <a href="https://edge.readthedocs.io/en/v1.1/index.html">https://edge.readthedocs.io/en/v1.1/index.html</a>. There is also a link to the documentation on the EDGE website. For basic information about EDGE, visit the <a href="https://edge.readthedocs.io/en/v1.1/EDGE_ABCs.html"> EDGE ABC’s</a>, that provide a brief overview of EDGE, the various workflows, and the computational environment restraints for local use.
</p>

<h3><a id="output" class="anchor" href="#output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output</h3>
<p>
  Once your project is completed, EDGE will create an output report that contains a brief summary of the run, along with a list of drop down menus that show the EDGE tools used and detailed reports. The final data is listed in these menus and visualized through a tool called Krona, an interactive browser that allows the exploration of hierarchical data with pie charts. You can also see the full data by downloading the report PDF's, found at the bottom of each output section.
</p>

<h3><a id="disscussions" class="anchor" href="#disscussions" aria-hidden="true"><span class="octicon octicon-link"></span></a>Discussions / Bugs Reporting</h3>

<p>We have created a mailing list for EDGE users. If you would like to recieve notifications about the updates and join the discussion, please join the mailing list by becoming the member of edge-users groups.</p>
<p><a href="https://groups.google.com/d/forum/edge-users">EDGE user's google group</a></p>

<p>We appreciate any feedback or concerns you may have about EDGE. If you encounter any bugs, you can report them to our GitHub issue tracker.</p>
<p><a href="https://github.com/LANL-Bioinformatics/EDGE/issues">Github issue tracker</a></p>

<h3><a id="citation" class="anchor" href="#citation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Publication</h3>

<strong>Enabling the democratization of the genomics revolution with a fully integrated web-based bioinformatics platform </strong>
Po-E Li; Chien-Chi Lo; Joseph J. Anderson; Karen W. Davenport; Kimberly A. Bishop-Lilly; Yan Xu; Sanaa Ahmed; Shihai Feng; Vishwesh P. Mokashi; Patrick S.G. Chain
<em>Nucleic Acids Research</em> 2016; <a href="http://nar.oxfordjournals.org/content/early/2016/11/28/nar.gkw1027.abstract">doi: 10.1093/nar/gkw1027</a>

<h3><a id="copyright" class="anchor" href="#copyright" aria-hidden="true"><span class="octicon octicon-link"></span></a>Copyright</h3>
<p>
Los Alamos National Security, LLC (LANS) owns the copyright to EDGE, which it identifies internally as LA-CC-14-007.  The license is GPLv3.  See <a href="http://www.gnu.org/licenses/gpl-3.0.en.html">LICENSE</a> for the full text.
</p>
<!--
<p>Having trouble with edge? Check out the documeentation at <a href="http://edge.readthedocs.io/en/latest">https://edge.readthedocs.org/en/latest/</a> or contact <a href="mailto:support@github.com">support@github.com</a> and we’ll help you sort it out.</p>
-->
      </section>
    </div>
    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer" style="background: #212121 url('images/los-alamos-lab-logo-white.png') no-repeat bottom right; ">
      <footer class="inner">
        <p class="copyright">Edge maintained by <a href="https://github.com/LANL-Bioinformatics">LANL-Bioinformatics</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
        <p>This project is funded by U.S. Defense Threat Reduction Agency [R-00059-12-0].</p>
      </footer>
    </div>

    

  </body>
</html>
